### PR TITLE
Improve performance of spellchecking checks

### LINKED ISSUE #2249: Investigate why the spellchecking checks are slow
https://github.com/PyCQA/astroid/issues/579#issuecomment-402068555 this comment mentioned that disabling the spellchecking checks improvement the runtime of pylint from half an hour to ~10 minutes, which suggests that something might be completely off with the spellchecking. We should investigate to see why is that, and if we can find that these are hurting the performance, we should either disable the spellchecking by default (as we do for the Py3k porting checks), or we should move them to an extension, so that they can be enabled explicitly by the user.

### PR DESCRIPTION
**Pull request by svmhdvn**
## Steps

- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [ ] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
This PR improves the performance of the spellchecking checks by 2x on average. It mainly does the following things:
- Replaces unmaintained `pyenchant` spellchecker with maintained (and faster) `hunspell` python bindings
- Adds a rudimentary english word tokenizer (a modified version of the tokenizer from `pyenchant`)
- Fixes minor bugs with ordering of the tokenizer's filters and chunkers
- Slightly refactors the `unittest_checker_spelling.py` tests to accomodate the new change
- Adds the option of placing tox positional arguments in a tox command (for better testing the spelling module locally and on CI)
- Adds `--spelling-dict-name` and `--spelling-dict-paths` options for users to select dictionary files (breaking change to config)

## Benchmarks
Datadog APM was used to benchmark the before and after of this PR. The corpus used was the entire `pylint/` source folder under this repo. These are some of the results taken from many samples:

### Before:

Flame graph of the function calls made during a run of the largest file in `pylint/`
image

Function call list
image

Pyenchant `suggest(word)` calls
image

Pyenchant `spellcheck(word)` calls
image

Pyenchant tokenizer `next()` calls
image

### After:

Flame graph of the function calls made during a run of the largest file in `pylint/`
image

Function call list
image

Hunspell `suggest(word)` calls
image

Hunspell `spellcheck(word)` calls
image

New pylint english word tokenizer `next()` calls
image

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
|  | :scroll: Docs |

## Related Issue

Closes #2249

### CODE REVIEW

**Comment by PCManticore** [comment]
@svmhdvn Great stuff! Thanks for working on this! Let me know once it's ready for review.

**Comment by svmhdvn** [comment]
Any update on this PR? I'd like to start fixing it based on any feedback you have :)

**Comment by PCManticore** [comment]
Hey @svmhdvn I'll review this weekend, thank you for waiting! I was swamped with other work and didn't have time for OSS.

**Comment by svmhdvn** [comment]
Hey @PCManticore, were you able to take a look at the changes?

**Comment by PCManticore** [comment]
@svmhdvn Not quite, sorry for the extensive delay! Will try to carve some time this week, please ping me if there'll be  no movement here.

File: examples/pylintrc [hunk 1]
```diff
@@ -354,13 +354,20 @@ min-similarity-lines=4
...
+spelling-dict-name=
 
-# Spelling dictionary name. Available dictionaries: none. To make it working
-# install python-enchant package.
-spelling-dict=
+# Paths string of directories where Hunspell dictionaries are installed.
+# To get a list of available paths, run `hunspell -D` in a shell and look for
+# the AVAILABLE DICTIONARIES section.
+# Pylint will search under these directories for the following files:
+#   <spelling-dict-name>.dic
```
Comment (PCManticore) [inline examples/pylintrc]: Is this something that can also be used on Windows for instance? while currently `pyenchant` is unmaintained, we still offer Windows support as long as the library is installable on that platform.

File: tox.ini [hunk 2]
```diff
@@ -6,8 +6,9 @@ skip_missing_interpreters = true
 deps =
    git+https://github.com/pycqa/astroid@master
    isort
-
-commands = pylint -rn --rcfile={toxinidir}/pylintrc --load-plugins=pylint.extensions.docparams, pylint.extensions.mccabe {envsitepackagesdir}/pylint
+; McCabe complexity checking is currently disabled in CI, too many warnings. Uncomment this line if desired.
```
Comment (PCManticore) [inline tox.ini]: Let's figure out why is that, we might need to refactor something either in this PR, or in `master` if the problem is coming from there.

File: tox.ini [hunk 3]
```diff
@@ -37,7 +38,7 @@ setenv =
     COVERAGE_FILE = {toxinidir}/.coverage.{envname}
 
 commands =
-    python -Wi {envsitepackagesdir}/coverage run -m pytest {envsitepackagesdir}/pylint/test/ {posargs:}
+    python -Wi {envsitepackagesdir}/coverage run -m pytest {posargs:} {envsitepackagesdir}/pylint/test/
```
Comment (PCManticore) [inline tox.ini]: Any reason for moving `posargs` around?

File: pylint/test/spelling/unittest_checker_spelling.py [hunk 4]
```diff
@@ -12,44 +12,44 @@
...
-    @set_config(spelling_dict=spell_dict)
-    def test_check_bad_coment(self):
+    def _config(self, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self.checker.config, key, value)
+        self.checker.open()
+        if not self.checker.initialized:
+            pytest.skip("Missing hunspell package or missing spelling dictionaries")
+
+    def test_check_bad_coment(self, dict_name, dict_paths):
```
Comment (PCManticore) [inline pylint/test/spelling/unittest_checker_spelling.py]: Any reason why we can't use `set_config` here instead? It feels weird that we're reinventing what `set_config` does, I assume just to use the pytest fixtures.

File: pylint/checkers/spelling.py [hunk 5]
```diff
@@ -23,132 +24,159 @@
...
+                return self.sub_tokens.pop()
+
+        return base_token
+
+    def tokenize(self, text):
+        self._text = text
+        self._offset = 0
+        return self
+
+    # pylint: disable=R0201
```
Comment (PCManticore) [inline pylint/checkers/spelling.py]: Why do we need this method? `pylint` runs on Python 3 so what is invoking it?

File: pylint/checkers/spelling.py [hunk 6]
```diff
@@ -23,132 +24,159 @@
...
+    # Chars to remove from start/end of words
+    strip_from_start = set("\"'`([<{")
+    strip_from_end = set("\"'`]).!,?;:>}")
+
+    def next(self):
+        while self._offset < len(self._text):
+            # Find start of next word
+            while self._offset < len(self._text) and self._text[self._offset].isspace():
+                self._offset += 1
+            sPos = self._offset
```
Comment (PCManticore) [inline pylint/checkers/spelling.py]: Some variable names in this function could be updated to use snake case

File: pylint/checkers/spelling.py [hunk 7]
```diff
@@ -233,48 +260,39 @@ def open(self):
...
-            filters=[
-                EmailFilter,
-                URLFilter,
-                WikiWordFilter,
-                WordsWithDigigtsFilter,
-                WordsWithUnderscores,
-                CamelCasedWord,
-                SphinxDirectives,
-            ],
+            filters=[EnglishWordFilter(), WikiWordFilter(), CamelCasedWordsFilter()],
```
Comment (PCManticore) [inline pylint/checkers/spelling.py]: Does hunspell automatically filters out the same tokens as the previous incantation, e.g. `URLFilter` and friends?

File: pylint/checkers/spelling.py [hunk 8]
```diff
@@ -283,73 +301,43 @@ def close(self):
...
+            col = m.regs[2][0]
+            indicator = (" " * col) + ("^" * len(word))
+
+            # Present up to N suggestions.
+            suggestions = ""
+            if self.config.max_spelling_suggestions > 0:
+                suggested_words = self.hobj.suggest(word)
+                del suggested_words[self.config.max_spelling_suggestions :]
+                suggestions = "Did you mean: '{}'?".format(
+                    "' or '".join(suggested_words)
```
Comment (PCManticore) [inline pylint/checkers/spelling.py]: No need to delete the suggested words at line 332, just use `"".join(suggested_words[:self.config.max_spelling_suggestions]`. I don't know why the former code was doing that as well.

File: pylint/checkers/spelling.py [hunk 9]
```diff
@@ -383,8 +371,6 @@ def visit_functiondef(self, node):
             return
         self._check_docstring(node)
 
-    visit_asyncfunctiondef = visit_functiondef
```
Comment (PCManticore) [inline pylint/checkers/spelling.py]: Is this intentional? This means we'll no longer analyze `AsyncFunctionDef` nodes.

**Comment by svmhdvn** [comment]
Awesome, thanks for your feedback! I'll get started on fixing the PR up.

**Comment by PCManticore** [comment]
Hey @svmhdvn Have you got a chance to continue this PR? Let me know if there's any feedback you need in the meantime.

**Comment by PCManticore** [comment]
Hey @svmhdvn Thanks again for the PR! Unfortunately it seems it still needs some work and we probably won't have time to get it. If you happen to have some time and want to continue this further, please do and we can reopen the PR when that happens.

**Pull request closed without merging**